### PR TITLE
Support array as template type bound.

### PR DESCRIPTION
**Pull request by drupol**
This PR:

- [x] Relates to https://github.com/phpstan/phpstan/issues/3931 
- [x] Has been inspired by https://github.com/phpstan/phpstan-src/commit/4e7d6c155c52e11f81bcabf75f84e14a2e0e1727

### CODE REVIEW

File: tests/PHPStan/Analyser/data/generics.php [hunk 1]
```diff
@@ -1487,3 +1487,76 @@ function newClassString($type): void
...
+	assertType('array(1 => true)', arrayBound1([1 => true]));
+	assertType("array('a', 'b', 'c')", arrayBound2(range('a', 'c')));
+	assertType('array<string>', arrayBound2([1, 2, 3]));
+	assertType('array(true, false, true)', arrayBound3([true, false, true]));
+	assertType("array(array('a' => 'a'), array('b' => 'b'), array('c' => 'c'))", arrayBound4([['a' => 'a'], ['b' => 'b'], ['c' => 'c']]));
+	assertType('array<string>', arrayBound5(range('a', 'c')));
+};
+
+/**
+ * @template T of array{0: string, 1: bool}git add
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/generics.php:1551]: `git add`

File: tests/PHPStan/Rules/Methods/CallMethodsRuleTest.php [hunk 2]
```diff
@@ -497,6 +497,15 @@ public function testCallMethods(): void
 				'Parameter #1 $date of method Test\HelloWorld3::sayHello() expects array<DateTime|DateTimeImmutable>|int, DateTimeInterface given.',
 				1732,
 			],
+			[
+				'Parameter #1 $a of method Test\InvalidReturnTypeUsingArrayTemplateTypeBound::bar() expects array<string>, array<int, int> given.',
+				1751,
+			],
+			[
+				'Unable to resolve the template type T in call to method Test\InvalidReturnTypeUsingArrayTemplateTypeBound::bar()
+    💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type',
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Methods/CallMethodsRuleTest.php:506]: The tip should be a third element of the array, without the bulb emoji

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
This looks solid! Please do these two changes, mark it as ready and I'll merge it :)

**Comment by drupol** [comment]
@ondrejmirtes If you want, we can wait a bit more and I'll submit the same kind of stuff for `iterable`, `Generator` and `Iterator`.

WDYT ?

File: tests/PHPStan/Analyser/data/generics.php (hunk 1 above)
Comment (drupol) [reply tests/PHPStan/Analyser/data/generics.php]: Fixed.

File: tests/PHPStan/Rules/Methods/CallMethodsRuleTest.php (hunk 2 above)
Comment (drupol) [reply tests/PHPStan/Rules/Methods/CallMethodsRuleTest.php]: Fixed!

**Comment by ondrejmirtes** [comment]
@drupol I think all of those are already supported.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by drupol** [comment]
Excellent! Thanks for the support into making this :)

**Comment by ZebulanStanphill** [comment]
Thanks for this PR. I think this is going to help quite a bit in one of the projects I'm working on.

**Pull request merged**